### PR TITLE
add HD-Painter pipeline

### PR DESCRIPTION
**Pull request by haikmanukyan**
# What does this PR do?

This PR adds [HD-Painter](https://github.com/Picsart-AI-Research/HD-Painter).
HD-Painter is a training-free approach for improving text-alignment of Stable Diffusion based text-guided inpainting models.
It  adds the `StableDiffusionHDPainterPipeline` which is an inpainting pipeline inheriting from `StableDiffusionInpaintPipeline`, 
as well as the `PAIntAAttnProcessor`, an attention processor implementing the PAIntA layer from the paper.

## Usage example

```python
import torch
from diffusers import DiffusionPipeline, DDIMScheduler
from diffusers.utils import load_image, make_image_grid

pipe = DiffusionPipeline.from_pretrained(
    "stabilityai/stable-diffusion-2-inpainting",
    custom_pipeline="hd_painter"
)
pipe.scheduler = DDIMScheduler.from_config(pipe.scheduler.config)

prompt = "wooden boat"
init_image = load_image("https://raw.githubusercontent.com/Picsart-AI-Research/HD-Painter/main/__assets__/samples/images/2.jpg")
mask_image = load_image("https://raw.githubusercontent.com/Picsart-AI-Research/HD-Painter/main/__assets__/samples/masks/2.png")

image = pipe (prompt, init_image, mask_image, use_rasg = True, use_painta = True, generator=torch.manual_seed(12345)).images[0]

make_image_grid([init_image, mask_image, image], rows=1, cols=3)
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

@sayakpaul

### CODE REVIEW

**Comment by sayakpaul** [comment]
Thanks for contributing!

We should be able to pass `custom_pipeline="..."` and not the way you're showing in the OP. You can refer to the community examples REAMDE to know more about it.

**Comment by haikmanukyan** [comment]
@sayakpaul

File: examples/community/README.md [hunk 1]
```diff
@@ -75,6 +76,48 @@ pipe = DiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", custo
...
...iV2HggzgdTAdm589tLz2J--6ebFly64rC0w0zBXjglZieCXzbKonH3N1ne-FHFQDCQrLKBvPuvfVRMHUz1JqBb_gJZTUdArrG9k84yh4ASVx46U4HGn0aIr9QXbFge39xGCmpohje9544TZpqr74mi7pD4l8Oi4qHc68RpRvvSyd2QDaF451RItQxXJmTTtJKpLPuy9R-HxMhnh-Z9EcrWFA2V4JA_lK8KTAkS_occ-T-K0GFsHtwEijEgNURjAljhFM9ciyGluOPaJIN9_jMtCL6bhbktBe16EsewNNlaDyqUWL9pCnrI0Gcpnh0lg899fEZdG_flfWUQkHTpFkQlOrNLZux7FBvWrUm3yMgvaeQMH4GPlrENBpnUvjz3gWCkeKeDQ0CFrEaXkbqYk16YKdiPk3dBmdlOTul0hVswI9vxDWWKz2zxgids_x0TjJmLQmLlXDKJgznMHKRDtjEvEKDwxQqG_tH9H7Os7oBj3GcJidi3eae-XNrEG0qg8lI1ogqc1IHU56IJdNUoz_WulFQ0Crvn6sOITiy5_X66rsqT130-mRVvKOirkLuCX1Ae9pIkiUOwtY1KYFe6F4cwRuoGbMbw9-1HCjF0r46QqBKsxh4RkWRcn4EJtX-3yPxCWuuCNHIZlDTfo1OzQWO3LVp_9XNTTR5DW0eEyLzt1yQU9OWxyWbk3CYTG-ONVBVA-JzIknpmpzJskxz5_sQRi3_iZYmca7lCRqgGUpcG7ONkUeThgIaItHob-6oYZbG6yjECG_F2I4weGi2cbD2-ooVaY8HgqLTX5XBSalSqF9hWtxPZvtH-laIbHYp2SMDmGOsE_T7Zi8g9veg_1rDq84nWg0IOkPGQFA9pv26hI-Rdt1aFnKqXhUNW0yNza07X-p-sDgEKr3vBRwPFKvjqfYGJPnBN5HWJzKxBbMgfX90DSQHvVIyiJiZ0hA_pyT-1xpIGbSJ_wg2xAPko_LkJWhPuBfiQlCE9JnTVkmGG-hj_ugWbWdp9tWMMbcleKBy68qtecX5HPy61jmvTLTOU6fVIKBR4Pc8wawkaSohm3la2P54viMO0mcVmOj4ps7YDlbai-Rtr7gxrcAlqyIvbjDpqanRHHO-4Vm8CuOQx3IzcYsHglE8Eyt6CJBCR057KoaiS3kYprWvkILI8znlFHpWvakLO6cUQSyqVDvsXefpOL)
```
Comment (sayakpaul) [inline examples/community/README.md:83]: Can you host this on a GitHub repo or maybe even in a repository under https://huggingface.co/PAIR organization and provide a link here?

File: examples/community/README.md [hunk 2]
```diff
@@ -75,6 +76,48 @@ pipe = DiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", custo
...
+Therefore, in this paper we introduce \textit{HD-Painter}, a completely \textbf{training-free} approach that \textbf{accurately follows to prompts} and coherently \textbf{scales to high-resolution} image inpainting.
+To this end, we design the \textit{Prompt-Aware Introverted Attention (PAIntA)} layer enhancing self-attention scores by prompt information and resulting in better text alignment generations.
+To further improve the prompt coherence we introduce the \textit{Reweighting Attention Score Guidance (RASG)} mechanism seamlessly integrating a post-hoc sampling strategy into general form of DDIM to prevent out-of-distribution latent shifts.
+Moreover, HD-Painter allows extension to larger scales by introducing a specialized super-resolution technique customized for inpainting, enabling the completion of missing regions in images of up to 2K resolution. 
+Our experiments demonstrate that HD-Painter surpasses existing state-of-the-art approaches qualitatively and quantitatively, achieving an impressive generation accuracy improvement of \textbf{61.4\%} vs \textbf{51.9\%}. 
+We will make the codes publicly available.
```
Comment (sayakpaul) [inline examples/community/README.md:94]: No Latex formatting on markdowns please.

**Comment by haikmanukyan** [comment]
@sayakpaul done

**Comment by sayakpaul** [comment]
Will merge once the CI is green.

**Comment by haikmanukyan** [comment]
@sayakpaul I was using wrong ruff version, it should pass the check now

**Comment by sayakpaul** [comment]
Thanks for your contributions.

**Pull request merged**